### ISSUE TITLE
can't use SD Card on LCD from marlin mode, but can use it from touchscreen mode

### ISSUE THREAD

**Issue by robrogers** [bigtreetech/BTT-TFT35-E3-V3.0#2]
can't use SD Card on LCD from marlin mode, but can use it from touchscreen mode

**Comment by hfog** [comment]
This seems like expected behavior. The Marlin mode is really just treating the TFT35 as a dumb replacement for the stock E3 screen so it doesn't have any access to extended features offered by the TFT. When in touch screen mode the TFT is really just a second device hooked up to the printer over serial, think just like Octoprint or Pronterface. So the touch menus are interacting with the TFT which in turn is driving the printer board over serial while the Marlin menus are interacting directly with the printer board.

**Comment by thisiskeithb** [comment]
Make sure you set `SDCARD_CONNECTION` to `LCD`:

https://github.com/MarlinFirmware/Marlin/blob/bd4e38860b0ceacd543df8d6cc03a252028efd56/Marlin/Configuration_adv.h#L1200-L1209

```cpp
/**
   * Set this option to one of the following (or the board's defaults apply):
   *
   *           LCD - Use the SD drive in the external LCD controller.
   *       ONBOARD - Use the SD drive on the control board. (No SD_DETECT_PIN. M21 to init.)
   *  CUSTOM_CABLE - Use a custom cable to access the SD (as defined in a pins file).
   *
   * :[ 'LCD', 'ONBOARD', 'CUSTOM_CABLE' ]
   */
#define SDCARD_CONNECTION LCD
```

**Comment by Mowhok** [comment]
I have the same issue just in reverse. in marlin mode i can go to sd card but in touch mode i get a unknwon command xyucal1 gco 

with any sdcard added

**Comment by oldman4U** [comment]
Hi Mowhok.

Could you please start a ticket using BTT TFT Touchscreen. Thank you

@robrogers. What is the status of this issue. Could you solve it in the meantime?

**Comment by oldman4U** [comment]
@bigtreetech Could you please close this unused ticket

This could help new users find tickets which already cover their topic.

Thank you

**Comment by fineobjetcs** [comment]
I have same troubles. 
I cant see SD in marlin mode and see it in BTT mode when i choose LSD in #define SDCARD_CONNECTION
But if i choose ONBOARD i see onboard SD in marlin mode and cant see it in BTT

**Comment by oldman4U** [comment]
In Marlin you can not see anything TFT related. In touch mode you can see the TFT and you can define if you want to see the sd card from the board or not in config.ini. For this the board sd card must be enabled in Marlin board firmware.

**Comment by Portzal** [comment]
Just to throw in my recent experience, I hooked up a new TFT70 and was unable to print from the SD card in Marlin mode, I switched to touch screen mode and was able to access the files on the SD card. 
Without changing the firmware on the board (SKR V1.4 Turbo) and connecting a TFT25 with hasn't had firmware update in over 8 months I was able to access the files on the SD card in Marlin mode.  
I have not tried updating the firmware on the TFT70, and will not be able to for the next 2 weeks. 
My opinion is it is a screen firmware issue. 
Old Screen with old firmware works as expected, new screen with new firmware doesn't.

**Comment by oldman4U** [comment]
@Portzal 

SD card of mainboard or TFT?

**Comment by Portzal** [comment]
@oldman4U screen.

**Comment by oldman4U** [comment]
Marlin can only see the sd card from the mainboard and not the sd card and the U disk from TFT. This is independent of the screen size as long as it is a TFT from BTT.

**Comment by Portzal** [comment]
@oldman4U I have a BTT TFT25 and a BTT TFT35 that both print from the SD card in Marlin simulation mode. The new TFT70 does not, so either there has been a hardware change with the new screen or it is a screen firmware issue.

**Comment by oldman4U** [comment]
But this has nothing to do with this ticket.

To be able to print from mainboard SD card using Marlin simulation the SD card functionality has to be enabled in the firmware of the mainboard (default OFF) and the TFT (default auto-detect - so off in case it is off in Marlin fw)

Hope this helps

![image](https://user-images.githubusercontent.com/54359396/93666223-ae935000-fa7c-11ea-8d04-8a2e6c324226.jpeg)

![image](https://user-images.githubusercontent.com/54359396/93666234-b6eb8b00-fa7c-11ea-8a3f-ab5db0334c77.jpeg)

**Comment by Portzal** [comment]
Forgot anything about the main board because I have not been referring to it. And from the way I read the original ticket neither is that. 
All my comments are about the BTT TFT screens and accessing the SD card inserted in the screen SD slot. Same as the originator of the ticket. 
With a SD card inserted in the BTT TFT70 I can access print files in the touch screen mode, if I switch to Marlin simulation mode on the TFT70 it does not show print files. But for example on my BTT TFT35 V3.0 with a SD card inserted in the screen SD slot I can access the files on the SD card and start a print.

**Comment by oldman4U** [comment]
Not sure we can leave the mainboard completely out of the game.

There is a setting in configuration_adv.h, which allows you to define which SD Card Marlin can see:
  _/**
   * Set this option to one of the following (or the board's defaults apply):
   *
   *           LCD - Use the SD drive in the external LCD controller.
   *       ONBOARD - Use the SD drive on the control board. (No SD_DETECT_PIN. M21 to init.)
   *  CUSTOM_CABLE - Use a custom cable to access the SD (as defined in a pins file).
   *
   * :[ 'LCD', 'ONBOARD', 'CUSTOM_CABLE' ]
   */
  #define SDCARD_CONNECTION LCD_

On my machine I use a SKR E3 DIP which is connected with a single EXP connector, and on this machine defining   #define SDCARD_CONNECTION LCD does not make any difference.

I could imagine, that using a mainboard with 2 EXP connectors could eventually work, but this is completely guesswork from my side.

So far I was never able to see the TFTs U Disk or SD card from the Marlin emulator, but only when using the Touchscreen mode. But I got a SKR v1.4 last week which I bought for testing. Maybe I can find the time to test it.

So when you use the SD Card of the TFT, does this look the same in Marlin like when you use the on board SD or different?

**Comment by Portzal** [comment]
EXP1 and EXP2 is connected to the SKR V1.4 Turbo from the BTT TFT35 V3.0, with a SD card inserted in the BTT TFT35 V3.0 SD card slot and With the BTT TFT35 V3.0 in Marlin simulation mode I can see the files on the SD card then select and print them. I have been using my screen in this way for over 8 months as I prefer Marlin simulation mode to the touch screen. 
No SD card is in the main board,

**Comment by Portzal** [comment]
> Not sure we can leave the mainboard completely out of the game.
> 
> 
> 
> There is a setting in configuration_adv.h, which allows you to define which SD Card Marlin can see:
> 
>   _/**
> 
>    * Set this option to one of the following (or the board's defaults apply):
> 
>    *
> 
>    *           LCD - Use the SD drive in the external LCD controller.
> 
>    *       ONBOARD - Use the SD drive on the control board. (No SD_DETECT_PIN. M21 to init.)
> 
>    *  CUSTOM_CABLE - Use a custom cable to access the SD (as defined in a pins file).
> 
>    *
> 
>    * :[ 'LCD', 'ONBOARD', 'CUSTOM_CABLE' ]
> 
>    */
> 
>   #define SDCARD_CONNECTION LCD_
> 
> 
> 
> On my machine I use a SKR E3 DIP which is connected with a single EXP connector, and on this machine defining   #define SDCARD_CONNECTION LCD does not make any difference.
> 
> 
> 
> I could imagine, that using a mainboard with 2 EXP connectors could eventually work, but this is completely guesswork from my side.
> 
> 
> 
> So far I was never able to see the TFTs U Disk or SD card from the Marlin emulator, but only when using the Touchscreen mode. But I got a SKR v1.4 last week which I bought for testing. Maybe I can find the time to test it.
> 
> 
> 
> So when you use the SD Card of the TFT, does this look the same in Marlin like when you use the on board SD or different?
> 
> 

I have not changed any of these settings in my Marlin build.

**Comment by oldman4U** [comment]
It is a miracle then...

**Comment by Portzal** [comment]
@oldman4U it's not a miracle it's just how it works. The inclusion of your comments in this thread have not been of any assistance as you are basing them on a setup that doesn't allow printing from the SD card in Marlin simulation mode. 
The fact is that you can access files from the SD Card inserted in the TFT screen slot in Marlin simulation mode when you have EXP1 and EXP2 connected to the main board.

**Comment by oldman4U** [comment]
@radek8

Sorry to bother you but you know the hard and software much better than me. 

What is needed to be able to print from the TFTs SD card, using MARLIN emulation mode?

Does this work on displays using just one EXP connection to the mainboard or does this only work with two?

Whats has to be activated/selected software wise?

Does this work on the new TFT70?

You help is appreciated.

THANK YOU

**Comment by Portzal** [comment]
As far as I know only SD card is supported in Marlin simulation mode, and again to my knowledge you need EXP1 and EXP2 cable's connected. I am back from holiday and can now do some fault finding of my own. I will flash the latest.bin from the Bigtreetech screen firmware to my TFT70 and see if the issue has been sorted. If that doesn't work I am going to load a previous firmware version that I have on my TFT35 V3.0 onto my new TFT70 and see if that resolves the issue.
Depending on what I find I will report back here and do a new bug report if needed.

**Comment by oldman4U** [comment]
Corrected my question to radek8.

Not sure the TFT70 will support older firmware than it has been released with. 😬

**Comment by radek8** [comment]
oldman4U to see SD card on TFT you have to meet 2 conditions.
1. You must be connected via EXP1 and EXP2 (+ correctly declared display type and pins of the EXP2 connector in Marlin). You write that you use EXP3, so you won't see SD on TFT.
2. You must be configured in Marlin
  #define SDCARD_CONNECTION LCD

As for the TFT70, I dont have it. He doesn't know if there are any changes on the TFT70 compared to the TFT35.

**Comment by radek8** [comment]
I would not record older FW due to SD support.
A better solution would be to lead the pins from the EXP2 connector (on the Marlin board) to an external SD card slot.

![image](https://user-images.githubusercontent.com/46979052/93708919-68042b00-fb3a-11ea-8f9a-c17bffc60edd.png)

**Comment by radek8** [comment]
for example, this
https://www.aliexpress.com/item/32813538365.html?spm=a2g0o.productlist.0.0.6f83443aItqWCY&algo_pvid=2d2c5e73-59fd-4d67-99c7-c8cea8132f53&algo_expid=2d2c5e73-59fd-4d67-99c7-c8cea8132f53-17&btsid=0b0a0ae216005941330491088e3213&ws_ab_test=searchweb0_0,searchweb201602_,searchweb201603_
,


https://www.aliexpress.com/item/32523546123.html?spm=a2g0o.detail.1000014.7.155625c6zYzQ8l&gps-id=pcDetailBottomMoreOtherSeller&scm=1007.14976.190542.0&scm_id=1007.14976.190542.0&scm-url=1007.14976.190542.0&pvid=f4a31b81-d134-4ad0-85f6-b9b7507523be&_t=gps-id:pcDetailBottomMoreOtherSeller,scm-url:1007.14976.190542.0,pvid:f4a31b81-d134-4ad0-85f6-b9b7507523be,tpp_buckets:668%230%23131923%2333_668%23808%234094%23149_668%23888%233325%235_4976%230%23190542%233_4976%232711%237538%23409_4976%233104%239652%231_4976%233141%239887%237_668%232846%238110%23382_668%232717%237559%2366_668%231000022185%231000066059%230_668%233422%2315392%23781_4452%230%23184418%230_4452%233474%2315675%23367_4452%233098%239599%23889_4452%233564%2316062%23560

**Comment by Portzal** [comment]
I'm out

**Comment by fineobjetcs** [comment]
Hi guys !

I find time for some testing. I removed my BTT 35 TFT from one printer and
installed it on another one with the same SKR1.4 and working display. I
make booth connection - rs232-tft and exp1 exp2

But i have not happened
LCD SD in Marlin mode still not worked

I tried onboard SD and it worked fine. I think it's BTT mistake because i
try to change all settings in Marlin and it doesn't matter

My mistake - that i bought this screen 👎because i can't use SD and U disk
in base (Marlin) mode, also i don't understand - for why Marlin mode on
this screen*,* without SD and U disk. Only for fan ?

**Comment by radek8** [comment]
SKR 1.4 has its own card slot and the TFT display also has its own card slot and as a bonus it has USB.
Isn't that enough for you?
Need to print from two cards at the same time?

You will never see USB from the TFT display in Merlin, it is not possible

**Comment by fineobjetcs** [comment]
Its enough for me while it will be work. But now not. I want to print
from one of two LCD slot SD or USB in MARLIN Mode.
Nobody cant say me how

**Comment by oldman4U** [comment]
OK, I also found some time for testing and it works using the SKR v1.4 and the TFT35 E3 v3. Using this hardware I can print from all 3 options using Touchscreenmode and using Marlin I can choose to print from mainboard SD card OR TFT SD card slot depending on the settings I defined.

@Portzal: Why have you deleted your comments? I wanted to thank you, because you have been right and I had the chance to learn something new - very cool! Thank you

@fineobjetcs Which version of the firmware are you using on your TFT? Can you send me the config.ini of it to check. Thank you

**Comment by fineobjetcs** [comment]
@oldman4U

I try last firmware for screen from BTT github/ and Marlin 2.0.5.3
Its last config but i try many changes for test

**Comment by oldman4U** [comment]
Marlin is currently 2.0.6.1 and seems to be OK.

TFT is called "master" and is currently 4 days old. There seems to be a problem with the precompiled versions but you will compile your own firmware anyway I guess;-)

**Comment by radek8** [comment]
So once again.
Marli only supports a single SD card!

when setting Marlin to
#define SDCARD_CONNECTION ONBOARD
There will be an SD card on board with Marlin available in Marlin.
An SD card connected via EXP2 on the display (even by touch TFT) in Marlin will not be available
Both cards + USB disk will be available in touch mode.


When setting Marlin to
#define SDCARD_CONNECTION LCD
The SD card on the Marlin board will be unavailable for both Marlin and touch mode.
The SD card on the display connected via EXP2 will be available in Marlin and in touch mode.
USB drive is only available in touch mode.

When connecting the display only via EXP3
the SD card in the display will not be available in Marlin with any settings.

Is my English clear enough?

**Comment by oldman4U** [comment]
To me everything is clear and I checked on the SKR v1.4 which I bought 2 weeks ago. The E3 DIP and Pro which i use work differently. Thank you

**Comment by radek8** [comment]
SKR E3 DIP has port EXP3, so you can't get to the SD card of the display from Marlin.

**Comment by radek8** [comment]
Of course, the settings of the display driver depend on this

Enable the driver for cables EXP1 and EXP2.
#define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
with it, the SD card pins are activated

You enable the driver for the EXP3 cable.
#define CR10_STOCKDISPLAY

**Comment by oldman4U** [comment]
Again. Thank you for your help. I really appreciate.

**Comment by oldman4U** [comment]
> SKR E3 DIP has port EXP3, so you can't get to the SD card of the display from Marlin.

Exactly. And the SKR Pro v1.1 and SD card support is a special chapter....🤪

**Comment by oldman4U** [comment]
> Of course, the settings of the display driver depend on this
> 
> 
> 
> Enable the driver for cables EXP1 and EXP2.
> 
> #define REPRAP_DISCOUNT_FULL_GRAPHIC_SMART_CONTROLLER
> 
> with it, the SD card pins are activated
> 
> 
> 
> You enable the driver for the EXP3 cable.
> 
> #define CR10_STOCKDISPLAY

Traffic light added this information to his READ ME BEFORE STARTING A TICKET because i asked him to do do. Maybe he can also add the information about the Sd support depending on the availability of EXP2.

**Comment by Portzal** [comment]
> OK, I also found some time for testing and it works using the SKR v1.4 and the TFT35 E3 v3. Using this hardware I can print from all 3 options using Touchscreenmode and using Marlin I can choose to print from mainboard SD card OR TFT SD card slot depending on the settings I defined.
> 
> 
> 
> @Portzal: Why have you deleted your comments? I wanted to thank you, because you have been right and I had the chance to learn something new - very cool! Thank you
> 
> 
> 
> @fineobjetcs Which version of the firmware are you using on your TFT? Can you send me the config.ini of it to check. Thank you

@oldman4U I deleted my comments from this thread because what I was saying was not being understood.
So I assume it would be pointless to have my comments here.

**Comment by oldman4U** [comment]
@Portzal 

But your comments pushed me into the right directions. So I am happy you are back. Sorry that I did not understand immediately. 

👍🏻😎

**Comment by fineobjetcs** [comment]
I have important news! I change screen from 3.5 to new 2.5 and SD start
working in marlin! With same settings in skr board. Thats troubles with 35
screen only. May be firmware or etc. 25 screen works with sd right and in
btt mode I see 3 medias: sd LCD, sd on board and u drive!

пн, 21 сент. 2020 г., 18:49 oldman4U <notifications@github.com>:

**Comment by oldman4U** [comment]
Using new TFT firmware this also works with TFT35. New TFT already comes with new firmware. I see all 3 in touchmode and one sd card in Marlin mode on my TFT35. 

Let me know in case you need help updating the TFT.

**Comment by RedCoder09** [comment]
So... find a solution yet? If so, what is it?

**Comment by oldman4U** [comment]
Are you robrogers or are you new? What is your mainboard and tft and which firmware are you using?

**Comment by oldman4U** [comment]
The solution is the latest firmware an the correct setup.

**Comment by Portzal** [comment]
I have updated my TFT70 V3.0 to latest firmware, still no good. I then compiled my own build altering settings that I thought may be interfering and created a .bin. I still get SD fail in Marlin simulation mode. I plug in my old TFT25 and no problem I can print from the SD card inserted in the screen in Marlin simulation mode. There is a bug with the new firmware.

**Comment by oldman4U** [comment]
Can you please attach your config.ini.

Thank you

**Comment by Portzal** [comment]
I am beginning to think it might be a hardware issue with my screen. I purchased it as soon as it was released and maybe poor quality. I have toggles all the relevant settings in config.ini, so I doubt me sharing it will help. I have tried it on a different board with exactly the same results.

**Comment by oldman4U** [comment]
Feel free to send me the config.ini or any other information which could help us to help you.

For Tony this worked very well today: https://github.com/bigtreetech/BIGTREETECH-TFT35-V3.0/issues/81

**Comment by Portzal** [comment]
I am going to cross reference the screen firmware .h files for the version I have on my TFT35 V3.0 with the latest version and see if I can pick anything up that may be causing the issue when I have a spare moment. 
As far as hardware possibilities it may be a bad pin in EXP1 or EXP2. But I would have to research the pins.
Seems like a good opportunity to start using the touch screen mode.

**Comment by oldman4U** [comment]
Ok. Maybe you can start your own ticket then and let us know👍🏻

@robrogers

Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by RedCoder09** [comment]
TFT35, SKR E3 Mini V1.2, Marlin 2.0, and I'm not robrogers.

**Comment by oldman4U** [comment]
Nice to meet you and nice hardware.

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by oldman4U** [comment]
Please help the community and close the ticket once you do not need it anymore. Thank you

**Comment by xdased14x** [comment]
i know this is an old ticket just checking that i understand.
i have ender 3 pro with skr mini e3 1.2, on marlin 2.03(will be updating to 2.0.7 shortly) i added tft 35 e3 v3 to it i can see sd card (on board only in sim mode) in touch i only see tft sd and u disk i want to use on board in both so i uncomment this   #define SDCARD_CONNECTION LCD in marlin config adv h(line 1166 on vscode)? i compiled my own code for the tft from the lastest on bbt tft code

**Comment by radek8** [comment]
line must be changed to
  #define SDCARD_CONNECTION ONBOARD

**Comment by xdased14x** [comment]
ok i'll try that thanks
so from lcd to onboard

**Comment by mark429** [comment]
> So once again.
> Marli only supports a single SD card!
> 
> when setting Marlin to
> #define SDCARD_CONNECTION ONBOARD
> There will be an SD card on board with Marlin available in Marlin.
> An SD card connected via EXP2 on the display (even by touch TFT) in Marlin will not be available
> Both cards + UDB disk will be available in touch mode.
> 
> When setting Marlin to
> #define SDCARD_CONNECTION LCD
> The SD card on the Marlin board will be unavailable for both Marlin and touch mode.
> The SD card on the display connected via EXP2 will be available in Marlin and in touch mode.
> USB drive is only available in touch mode.
> 
> When connecting the display only via EXP3
> the SD card in the display will not be available in Marlin with any settings.
> 
> Is my English clear enough?

Thank you for this. I didn't understand why it would work from SKR1.3 but not SKR Mini1.2,  but now it makes sense. 
Thank you!

**Comment by radek8** [comment]
similar query
https://github.com/bigtreetech/BIGTREETECH-TouchScreenFirmware/issues/1542
